src/mito2/src/row_converter.rs

```diff
@@ -0,0 +1,409 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::Buf;
+use common_base::bytes::Bytes;
+use common_time::time::Time;
+use common_time::{Date, Interval, Timestamp};
+use datatypes::data_type::ConcreteDataType;
+use datatypes::prelude::Value;
+use datatypes::value::ValueRef;
+use memcomparable::{Deserializer, Serializer};
+use paste::paste;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, ResultExt};
+
+use crate::error;
+use crate::error::{
+    FieldTypeMismatchSnafu, NotSupportedFieldSnafu, RowLengthMismatchSnafu, SerializeFieldSnafu,
+};
+
+/// Row value encoder/decoder.
+pub trait RowCodec {
+    /// Encodes rows to bytes.
+    fn encode<'a, I>(&self, rows: I) -> error::Result<Vec<u8>>
+    where
+        I: Iterator<Item = &'a [ValueRef<'a>]>;
+
+    /// Decode row values from bytes.
+    fn decode(&self, bytes: &[u8]) -> error::Result<Vec<Vec<Value>>>;
+}
+
+pub struct SortField {
+    data_type: ConcreteDataType,
+}
+
+impl SortField {
+    pub fn new(data_type: ConcreteDataType) -> Self {
+        Self { data_type }
+    }
+
+    pub fn estimated_size(&self) -> usize {
+        match &self.data_type {
+            ConcreteDataType::Boolean(_) => 2,
+            ConcreteDataType::Int8(_) | ConcreteDataType::UInt8(_) => 2,
+            ConcreteDataType::Int16(_) | ConcreteDataType::UInt16(_) => 3,
+            ConcreteDataType::Int32(_) | ConcreteDataType::UInt32(_) => 5,
+            ConcreteDataType::Int64(_) | ConcreteDataType::UInt64(_) => 9,
+            ConcreteDataType::Float32(_) => 5,
+            ConcreteDataType::Float64(_) => 9,
+            ConcreteDataType::Binary(_) => 11,
+            ConcreteDataType::String(_) => 11, // a non-empty string takes at least 11 bytes.
+            ConcreteDataType::Date(_) => 5,
+            ConcreteDataType::DateTime(_) => 9,
+            ConcreteDataType::Timestamp(_) => 10,
+            ConcreteDataType::Time(_) => 10,
+            ConcreteDataType::Interval(_) => 18,
+            ConcreteDataType::Null(_)
+            | ConcreteDataType::List(_)
+            | ConcreteDataType::Dictionary(_) => 0,
+        }
+    }
+}
+
+impl SortField {
+    fn serialize(
+        &self,
+        serializer: &mut Serializer<&mut Vec<u8>>,
+        value: &ValueRef,
+    ) -> error::Result<()> {
```

Comment:
question: will the serializer encode one more byte to represent whether the value is nullable?

Reply:
> question: will the serializer encode one more byte to represent whether the value is nullable?

An extra bytes will be encoded before actual value to represent if is null